Description: Try to reconnect HSP SCO.
 This is usually because an error occurred when the SCO connection
 was initiated, we need to try to reconnect to optimize the
 handling of this problem.
Author: Chengyi Zhao <zhaochengyi@uniontech.com>

Origin: https://gerrit.uniontech.com/plugins/gitiles/base/pulseaudio/+/a9bf43439ded8bd2d1d63aaea0d43ff56d83ba76
Bug: https://pms.uniontech.com/zentao/bug-view-88648.html
Last-Update: 2022-05-25

--- pulseaudio-15.0+dfsg1.orig/src/modules/bluetooth/backend-native.c
+++ pulseaudio-15.0+dfsg1/src/modules/bluetooth/backend-native.c
@@ -104,6 +104,8 @@ static uint32_t hfp_features =
  * The choice seems to be a bit arbitrary -- it looks like at least channels 2, 4 and 5 also work*/
 #define HSP_HS_DEFAULT_CHANNEL  3
 
+#define SCO_RECONNECTION_COUNT  3
+
 #define PROFILE_INTROSPECT_XML                                          \
     DBUS_INTROSPECT_1_0_XML_DOCTYPE_DECL_NODE                           \
     "<node>"                                                            \
@@ -316,12 +318,22 @@ fail:
 static int sco_acquire_cb(pa_bluetooth_transport *t, bool optional, size_t *imtu, size_t *omtu) {
     int sock;
     socklen_t len;
+    int i;
 
     if (optional)
         sock = sco_do_accept(t);
-    else
-        sock = sco_do_connect(t);
-
+    else {
+        for (i = 0; i < SCO_RECONNECTION_COUNT; i++) {
+            sock = sco_do_connect(t);
+
+            if (sock < 0) {
+                pa_log_error("err is %s and reconnection count is %d", pa_cstrerror(errno), i);
+                pa_msleep(300);
+                continue;
+            } else
+                break;
+        }
+    }
     if (sock < 0)
         goto fail;
 
